client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -144,27 +145,36 @@ private static <T> Iterator<T> flatten(Iterator<Iterable<T>> iterator, long maxR
         return stream.iterator();
     }
 
-    private static class AsyncIterator<T>
+    @VisibleForTesting
+    static class AsyncIterator<T>
             extends AbstractIterator<T>
     {
         private static final int MAX_QUEUED_ROWS = 50_000;
         private static final ExecutorService executorService = newCachedThreadPool(
                 new ThreadFactoryBuilder().setNameFormat("Trino JDBC worker-%s").setDaemon(true).build());
 
         private final StatementClient client;
-        private final BlockingQueue<T> rowQueue = new ArrayBlockingQueue<>(MAX_QUEUED_ROWS);
+        private final BlockingQueue<T> rowQueue;
         // Semaphore to indicate that some data is ready.
         // Each permit represents a row of data (or that the underlying iterator is exhausted).
         private final Semaphore semaphore = new Semaphore(0);
-        private final CompletableFuture<Void> future;
+        private final Future<?> future;
+        private volatile boolean finished;
 
         public AsyncIterator(Iterator<T> dataIterator, StatementClient client)
+        {
+            this(dataIterator, client, Optional.empty());
+        }
+
+        AsyncIterator(Iterator<T> dataIterator, StatementClient client, Optional<BlockingQueue<T>> queue)
         {
             requireNonNull(dataIterator, "dataIterator is null");
             this.client = client;
-            this.future = CompletableFuture.runAsync(() -> {
+            this.rowQueue = queue.orElseGet(() -> new ArrayBlockingQueue<>(MAX_QUEUED_ROWS));
+            this.finished = false;
+            this.future = executorService.submit(() -> {
                 try {
-                    while (dataIterator.hasNext()) {
+                    while (!Thread.currentThread().isInterrupted() && dataIterator.hasNext()) {
```

Comment:
`Thread.currentThread().isInterrupted()` is unreliable (interruption may be swallowed)
also, this clears interruption flag.

why not `while (!finished ..` ?